sinonjs/sinon
# Fix #1274 callCount property is incorrect

#### Purpose (TL;DR)
Fix issue #1274, Rewrite matchingFake in spy.js

#### Background (Problem in detail)
`spy.withArgs` function makes fake object and store `this.fakes`, depending on the combination of arguments.
And `spy.invoke` function makes use of `this.fakes`.

`matchingFake` function called by`spy.invoke`, returns the only one fake if arguments are matched.
And `spy.invoke` update the fake properties like `callCount`.
In this time, if some fakes in `this.fakes` are matched arguments, It make this issue happen.

For example
```
spy.withArgs(1, 1);
spy.withArgs(1); // spy.fakes = [ fake(1, 1), fake(1) ]
spy.invoke(1, 1); // invoke increment callCount either fake(1) or fake(1, 1), but not both.
```

#### Solution
Modify matchingFake to matchingFakes, and return array of all fakes matched.
`invoke` function updates properties of this and all matched fakes.
`spy.withArgs` and `stub` are affected by the above fix.
`spy.withArgs` Fixed to fit conventional working.


#### Additional
I investigate around the `matchingFake` function like stub.
And found case3 of #1274.
Conventionally, matchingFake return last fake (pop()).
I Fixed to fit conventional working, but case3 is not solved.
So, I commit 4e68e4d to make use of the most matched argument.

#### How to verify
1. Check out this branch (see github instructions below)
2. `npm install`
3. `npm test` # include test for this issue
